### ISSUE TITLE
Slow Transfer Speeds via socket slcan

### ISSUE THREAD

**Issue by yannick9906** [meatpiHQ/wican-fw#204]
Hey there,

my college just wrote on discord.
We try to use the WiCAN to flash ECUs via UDS Transfer Data. Our current development tool uses python-can to establish a socket connection with the WiCAN.
Previously we used a J2534 Interface and got a rate of about 70seconds for the 1.2MB flash file. With WiCAN it takes about 980 seconds, which is quite long compared to my arduino slcan test board (that one needs 400 seconds with 115200baud)
We're using the latest (3.30) firmware, WiCAN is in AP only mode, Task Manager show a datarate of 48kbits.
We also looked into it via wireshark and saw that the WiCAN response for the isotp frame takes about 57ms and our next frame follows after just 1ms.

To be fair, we are not fixed on using slcan, if there is another protocol to be easily used with python-can we are also open to use another one.

**Comment by meatpiHQ** [comment]
@yannick9906 I believe I may have found the issue. I’ll send you a firmware update later to resolve it.

**Comment by meatpiHQ** [comment]
@yannick9906 Please try this version. This should improve things.


> Previously we used a J2534 Interface and got a rate of about 70seconds for the 1.2MB flash file. 

Is this using USB/Serial or TCP?

[wican-fw_v334_hv300.zip](https://github.com/user-attachments/files/17029316/wican-fw_v334_hv300.zip)

**Comment by yannick9906** [comment]
The J2534 interface is via USB.
We flashed the firmware and get now a response a lot quicker. Sadly I couldn't test the speed right away, since python's slcan implementation actually can't handle that many messages at once and fails. I had this issue already with higher baudrate for my arduino test board. I'll look into that later, but it looks like your fix has worked as it should.

In the meantime we started on implementing the RealDash protocol (to take advantage of the lesser overhead needed) and got basic communication working. As soon as we tried hitting it with the flash we faced the issue, that your current RealDash parsing can only handle one message within a tcp frame. But tcp automatically combines multiple messages into a single frame and so does WiCAN for sending as well. Should I open another issue for this too?

**Comment by meatpiHQ** [comment]
@yannick9906 I believe it would be better to implement ISOTP support in the firmware, as it's more suitable for this type of application compared to using the RealDash protocol. I’ll send you a new firmware to test later.

**Comment by yannick9906** [comment]
@meatpiHQ Thank you for that great support, ISOTP would indeed be a very nice direct solution. Do you already have some sort of protocol in mind?

**Comment by meatpiHQ** [comment]
@yannick9906 I couldn't find a protocol that efficiently handles the transfer. The bottleneck is the CAN bus, which limits the TCP transfer speed. WiCAN can buffer a few kilobytes of data, but this means the buffer will constantly overflow. I believe the best solution is to define an efficient but simple protocol that encapsulates multiple CAN frames. It should send an acknowledgment once the transmission of all encapsulated frames to the CAN bus is completed, allowing the next chunk to be sent. Let me know if If you have a preferred protocol that might be suitable for this application. 

I can send you test code to do this fairly quickly. I'm interested in such an implementation because I have many users who are very interested in this application.

**Comment by yannick9906** [comment]
@meatpiHQ To make it easy, we could add another frame type to the RealDash Protocol:

```
FRAME 99:
4 Bytes Header: 0x99, 0x33, 0x22, 0x11
2 Bytes       : ISO-TP Length
# Bytes       : Payload (up to 4095 as of ISO-TP spec)
```

Checksum should not be needed since TCP already takes care of that
I think it should be the best performing one, since it is already a binary protocol and has not much overhead.
The acknowledgment will be handled by a positive response from the ECU for the iso-tp frame

**Comment by meatpiHQ** [comment]
@yannick9906 how do you propose to set the CAN ID?

**Comment by yannick9906** [comment]
@meatpiHQ I actually forgot it, i updated my message above :)

**Comment by meatpiHQ** [comment]
@yannick9906 please try this new firmware version. Make sure to select the new protocol "frame99".

I've also tested with small python code, it expect an ack frame to continue the transfer  ``` b'993322110000010100080000000000000000' ```

I'm curious to know the transfer speed!


<img width="202" alt="image" src="https://github.com/user-attachments/assets/fa495b49-6d01-4be0-b6f0-68956d508c58">


[wican-fw_v335_hv300.zip](https://github.com/user-attachments/files/17072763/wican-fw_v335_hv300.zip)


```
import socket
import time
import struct

# Configuration
TCP_IP = '192.168.0.83'
TCP_PORT = 3333
BUFFER_SIZE = 4096  
DELAY_BETWEEN_CHUNKS = 0.001  
FRAME_HEADER = b'\x99\x33\x22\x11'  
IDENTIFIER = 0x12345678  
ACK_IDENTIFIER = 0x0000000101  
ACK_PAYLOAD_SIZE = 8  
ACK_EXPECTED_PAYLOAD = b'\x00' * ACK_PAYLOAD_SIZE

file_path = 'test.bin'

def create_frame_99(identifier, payload):

    header = FRAME_HEADER
    identifier_bytes = struct.pack('>I', identifier)  
    payload_length = struct.pack('>H', len(payload))  

    frame = header + identifier_bytes + payload_length + payload
    return frame

def wait_for_ack(sock):

    try:
        ack_frame = sock.recv(18)
        print("ack_frame")
        print(ack_frame)
        if len(ack_frame) < 18:
            raise ValueError("Incomplete ACK frame received")
        
        header = ack_frame[:4]
        identifier, = struct.unpack('>I', ack_frame[4:8])
        payload_length, = struct.unpack('>H', ack_frame[8:10])
        payload = ack_frame[10:]
        print("identifier:")
        print(identifier)

        if header == FRAME_HEADER and identifier == ACK_IDENTIFIER and payload == ACK_EXPECTED_PAYLOAD:
            print("ACK received")
            return True
        else:
            print(f"Invalid ACK frame: {ack_frame}")
            return False

    except socket.error as e:
        print(f"Socket error while waiting for ACK: {e}")
        return False

def send_file_over_tcp():
    time.sleep(2)
    try:
        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
            s.connect((TCP_IP, TCP_PORT))
            print(f"Connected to {TCP_IP}:{TCP_PORT}")

            with open(file_path, 'rb') as f:
                while True:
                    chunk = f.read(BUFFER_SIZE)
                    if not chunk:
                        # EOF reached
                        break

                    frame = create_frame_99(IDENTIFIER, chunk)

                    s.sendall(frame)
                    print(f"Sent {len(chunk)} bytes as Frame 99")

                    if not wait_for_ack(s):
                        print("Failed to receive valid ACK, retrying...")
                        break 

                    time.sleep(DELAY_BETWEEN_CHUNKS)

            print("File transfer completed successfully")

    except FileNotFoundError:
        print(f"File not found: {file_path}")
    except socket.error as e:
        print(f"Socket error: {e}")
    except Exception as e:
        print(f"An error occurred: {e}")

if __name__ == "__main__":
    send_file_over_tcp()

**Comment by yannick9906** [comment]
Hey, firstof thank you for your great work. I tried to update the wican, but the firmware version stays on the 3.34 i had previously and no option for frame99 is available

**Comment by meatpiHQ** [comment]
@yannick9906 can you please reboot and try again multiple times. You're the 2nd user this week who reported this issue, I still cannot reproduce it. Which browser are you using? Does it return any error?

**Comment by yannick9906** [comment]
I am using Arc for Windows as Browser (Chromium). After a while I get a timeout for the request in the browser and the yellow message light won't show anything until I powercycle.
Edit: I checked in DevTools and the ota.bin POST fails with a ERR_CONNECTION_RESET. If I want to reload the page, the browser asks if I want to resubmit the form, but it fails again and again with the same error

**Comment by meatpiHQ** [comment]
@yannick9906 can you try with a different browser? 

If all else fails you might have to open the device and flash it through a USB cable. If so, you can use the other files mentioned in the link below from the latest firmware update, those files have not changed.

https://meatpihq.github.io/wican-fw/config/firmware-update#_2-usb-flash

**Comment by yannick9906** [comment]
@meatpiHQ I also tried with edge and plain chrome neither work. I also tried with my phone and get a text error message "Failed to write file to storage". If I can help you with some more information maybe via USB to give you some more info, i can do that before I flash via usb

**Comment by meatpiHQ** [comment]
@yannick9906 the error "Failed to write file to storage" is usually because you're uploading the zip file and not the binary inside. Please unzip first and upload the bin.

**Comment by yannick9906** [comment]
@meatpiHQ indeed i was uploading the zip file. But i did so also when updating to 3.34. But i'll now upload the binary, since that is working :)

**Comment by yannick9906** [comment]
@meatpiHQ I tested the firmware, and it accepts my 99 frame, but not as intended:
If i send only 7 or less bytes (the maximum in a Single Frame from ISOTP) it would be nice if it could be padded with 0x00 or 0xAA to fill 8 bytes can frame.
If i send more that 8 bytes it just outputs one frame to the canbus.

i tried:
`send: 99332211 000007e0 0007 3e 00 aa aa aa aa aa`
and got on canbus:
`7e0 <7>: 3e 00 aa aa aa aa aa`
but was expecting:
`7e0 <8>: 02 3e 00 aa aa aa aa aa` aka ISO TP Single Frame

**Comment by yannick9906** [comment]
Thank you for the quick response, it fill up the frame now, but still outputs only the raw can frame and not the ISO TP structure around it.
It should prepend 0x02 to the message, meaning 0 for "Single Frame" and 2 for "2 Bytes Data"

**Comment by meatpiHQ** [comment]
@yannick9906 please double check the format, im still working through the protocol 

[wican-fw_v337_hv300.zip](https://github.com/user-attachments/files/17078783/wican-fw_v337_hv300.zip)

**Comment by yannick9906** [comment]
Now I won't get any output on the canbus...

Also to note:
maybe we also need to define can_ids that use iso-tp. Since the ecu I am testing with only uses iso-tp on the diagnosic IDs 0x7e0 and 0x7e8, the rest are just normal can frames that could be filtered or send via 44 Frame

**Comment by meatpiHQ** [comment]
@yannick9906 how do you suggest we set the CAN ids, to make it generic?

**Comment by meatpiHQ** [comment]
We can make a configuration frame 
```
4 Bytes Header: 0x99, 0x33, 0x22, **0x01**
4 Bytes       : TX Identifier
4 Bytes       : RX Identifier
```

**Comment by yannick9906** [comment]
the can bitrate would also be nice. Apart from that maybe the byte used for frame padding could also be defined.

If you need a quite comprehensive article on iso-tp (or can-tp) you can read this:
https://piembsystech.com/can-tp-protocol/
That's where I learned iso-tp the first time.

**Comment by yannick9906** [comment]
@meatpiHQ Thank you once again for that crazy work!!

I am implementing our connector currently and just encountered a small issue: When sending a ISOTP message with less that 7 bytes the remaining can frame should be filled with the padding byte:
`ISO-TP Data: 3E 00, length: 2`
`CAN Frame: 02 3E 00 AA AA AA AA AA. length: 8`
Frames longer than 7 seem to work just fine

**Comment by meatpiHQ** [comment]
@yannick9906 No worries! Please try this version. 

[wican-fw_v339_hv300.zip](https://github.com/user-attachments/files/17096473/wican-fw_v339_hv300.zip)

**Comment by yannick9906** [comment]
@meatpiHQ still only `02 3E 00 `(length 3 byte) on the canbus

**Comment by meatpiHQ** [comment]
@yannick9906 ah forgot to set the DLC to 8.

[wican-fw_v340_hv300.zip](https://github.com/user-attachments/files/17096640/wican-fw_v340_hv300.zip)

**Comment by yannick9906** [comment]
@meatpiHQ Thank you for the quick fix. That works now.
I was wondering why i still got no answer from the ecu and I noticed the rx and tx IDs might be swapped?
I send this config:
Rx: 7e8, Tx: 7e0
`<FRAME99> send raw: COMMAND 04    0000 07e8 0000 07e0`
transmission on canbus happens with ID 7e8, but should be 7e0

**Comment by meatpiHQ** [comment]
@yannick9906 Does it work if you swap the ID's?

**Comment by yannick9906** [comment]
@meatpiHQ I swapped the IDs and the ecu answered. But the answer also was a ISO-TP message and I only got a raw receive for that message

**Comment by meatpiHQ** [comment]
@yannick9906 Yes only raw is supported for now. I will define a new response for the ISO-TP frame received. 

The ISO-TP frame received CAN ID will be 7e8 right?

**Comment by yannick9906** [comment]
@meatpiHQ Yes, we use 0x7e0 for sending, 0x7e8 for receiving. Or basically the value set by command 0x04 as RX ID

**Comment by meatpiHQ** [comment]
@yannick9906 please try the new firmware, I've added a new frame to receive iso-tp from CAN. i also updated the test code.

Let me know how it goes!

[wican-fw_v341_hv300.zip](https://github.com/user-attachments/files/17100555/wican-fw_v341_hv300.zip)
[frame_99_transfer_test.zip](https://github.com/user-attachments/files/17100566/frame_99_transfer_test.zip)

#### Structure
| Field         | Size  | Description                                      |
|---------------|-------|--------------------------------------------------|
| Header        | 3     | Constant header: `0x99, 0x33, 0x22`              |
| Command       | 1     | Command byte: `0x10`                             |
| ISO-TP Length | 2     | Length of the payload, as per ISO-TP specification |
| Payload       | Varies| Payload, up to 4095 bytes                        |

---

**Comment by meatpiHQ** [comment]
@yannick9906 my tests shows it's much faster than that. The esp should be able to handle it.

Is there a reason you're sending 128 byte frame? Why not the full 4095 bytes?

One reason for the slow transfer could be the 0 block size, it's possible I'm handling that incorrectly. All my test were done on block size 255.

**Comment by meatpiHQ** [comment]
@yannick9906 according to my test yesterday I transferred 1.5Mb file in 104 seconds the payload size of each frame_99 is 4095. Can you share your script so I can try to reproduce the issue here?

> Does the ACK for the large frame come, after the complete frame has been sent on the canbus or after it has been just received?

The ACk comes after all blocks has been transferred and for the last block is put into the CAN transmit buffer, so after receiving the ACK you can send the next frame directly or wait for any frame you expect from the ECU. 

> the rx/tx ids are still swapped?

Yes, but this is a minor issue. Easy fix.

**Comment by yannick9906** [comment]
Hey @meatpiHQ,
we sadly need to keep 128 Bytes, since that is what the ecu can handle. Some newer ones do it with 256 Bytes, but all far from the 4095 Bytes, mostly because their CAN Buffers aren't that big.
We also can't change what flow control we get from the ECU and we do need to wait for the ECU response, since it could always be a "21 Busy Repeat Request"

**Comment by meatpiHQ** [comment]
@yannick9906 in these case TCP will never be as fast USB, at least to my knowledge, the flow control is not an issue I've just checked.

Do you have any delays in your code after receiving the ACK?

The other option would be uploading the whole to flash memory and implement the ECU flashing logic on the ESP32 directly.

**Comment by yannick9906** [comment]
@meatpiHQ I also think implementing at least the logic for the large file transfer would deliver the best results. As soon as you update the branch for the frame99 protocol, I would try implementing that logic by myself, since it requires at lot specifics.

But nonetheless the frame99 protocol ensures a much more stable connection, when ISO-TP is needed or required by the ECU for the communication that is less data heavy.

Regarding the delays in my Code, from what Wireshark told me there were only 0.0003 sec between the Frame from WiCAN with the ECU Response and Frame to WiCAN with the next data.
I sadly can't measure the delays between the last ISO-TP Frame and the ECU response on the CAN Bus, since my other CAN receiver doesn't send timestamps, so I can't check if the ECU needs the time or the processing in the ESP32. Or maybe the TCP Stack in the ESP32?
In python at least I could set this option to tell TCP to directly send new data and not wait to fill a frame:
`self.client_socket.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)`
Maybe we can set this in the ESP32 as well?

**Comment by meatpiHQ** [comment]
@yannick9906 I've pushed the latest to https://github.com/meatpiHQ/wican-fw/tree/new-protocol

> Regarding the delays in my Code, from what Wireshark told me there were only 0.0003 sec between the Frame from WiCAN with the ECU Response and Frame to WiCAN with the next data.

Try removing all delays, this helped in my python script.

> . Or maybe the TCP Stack in the ESP32

I think it's possible to optimize the code further.

> self.client_socket.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
> Maybe we can set this in the ESP32 as well?

This flag TCP_NODELAY lwIP is available in ESP32 lwIP. I did not try this yet.

**Comment by meatpiHQ** [comment]
@yannick9906 I believe there’s a solution for this issue, how about sending multiple iso-tp frames of size 128 bytes in a single frame99.

I can introduce another FRAME99 command where you can specify the size of the ISO-TP frames. For example, you could send a FRAME99 with 4095 bytes and specify 128-byte iso-tp frames, and WiCAN can handle the transfer in 128-byte chunks, responding with an frame99 ACK once all the transfer is complete. This should help overcome TCP latency.

Let me know what you think of this.


#### Structure
| Field         | Size  | Description                                      |
|---------------|-------|--------------------------------------------------|
| Header        | 3     | Constant header: `0x99, 0x33, 0x22`              |
| Command       | 1     | Command byte: `0x13`                             |
| Frame99 Payload Length| 2| Length of the payload|
| ISO-TP Length | 2     | Length ISO-TP frames |
| Payload       | Varies| Payload, up to 4095 bytes                        |

**Comment by yannick9906** [comment]
@meatpiHQ I implemented a proof of concept for that already. Since it is a higher level communication than ISO-TP itself it needs to also handle the ECU responses like "busy repeat request" and sequence numbers of the higher level messages.

What we also noticed: When using WiCAN in the vehicle with a much "fuller" CAN-Bus I often get timeouts for some Frame99 messages. (I yet need to check if the python TCP implementation misses the messages or if wican doesn't answer) It would be nice if we could filter the messages to only get messages for the ECU Rx Tx IDs set.

**Comment by meatpiHQ** [comment]
@yannick9906 You can setup acceptance filters in the esp32 CAN controller. This is basically a hardware filter that would only allow an ID or range CAN IDs to pass.

https://docs.espressif.com/projects/esp-idf/en/v5.2.3/esp32c3/api-reference/peripherals/twai.html?highlight=twai_filter_config_t#_CPPv420twai_filter_config_t

**Comment by yannick9906** [comment]
@meatpiHQ I attached my version of the frame_99.c, maybe you can have a look: [frame_99.zip](https://github.com/user-attachments/files/17904527/frame_99.zip)

- I added the frame_99 command `0xA0` for the Large Transfer
- the `0x04` setECUTxRxIDs now also sets twai filters
- Since I want to use WiCAN frame99 to still support RAW CAN, I added the command `0x09` enableISOTP.

The `0xA0` command works now as it should and we get around 19 - 20 kbits transfer speeds, which is the limit of the ECU itself.
But sometimes get register traces while waiting for the ECU answers with xQueueReceive:

![image](https://github.com/user-attachments/assets/27130f5b-c145-4d87-918e-cc16158ece60)

This happens after following:
- WiCAN sends the first 2 128 Byte Blocks to the ECU
- The third Block receives a negative Response "Busy Repeat Request"
- so WiCAN retries this block until it gets a positve response.
- In this example it fails while waiting for the FC frame with xQueueReceive
This happens consistently with the amount of load on the canbus. The higher the load, the more often this happens

**Comment by meatpiHQ** [comment]
@yannick9906 the first thing you should try is to disable the log output

`esp_log_level_set("*", ESP_LOG_NONE);`

**Comment by yannick9906** [comment]
@meatpiHQ I tried it already. I get the same register dump and also no answer for the frame_99 `0xA0` message

**Comment by meatpiHQ** [comment]
@yannick9906 It's better to keep the logging output to minimum. It will severely affect the performance. Please try the following.  

1- Try increasing the stack of the task calling `frame_99_parse_data`, maybe double check the task watermark.
2- Try this receive_flow_control_frame. You should let `xQueueReceive` block for sometime

```
static frame_99_err_t receive_flow_control_frame(uint32_t identifier, bool is_extended, flow_control_params_t *fc_params)
{
    twai_message_t rx_message;
    TickType_t start_time = xTaskGetTickCount();
    const TickType_t timeout = pdMS_TO_TICKS(1000); // 1 second timeout
    TickType_t remaining_time;

    while ((remaining_time = timeout - (xTaskGetTickCount() - start_time)) > 0)
    {
        // Check if a message is received, waiting for remaining time
        if (xQueueReceive(frame_99_can_rx_queue, &rx_message, remaining_time) == pdPASS)
        {
            // Filter messages based on identifier (use Tx ID) and frame type
            if (rx_message.identifier == identifier &&
                rx_message.extd == (is_extended ? 1 : 0))
            {
                uint8_t pci_type = rx_message.data[0] & 0xF0;

                if (pci_type == 0x30)
                {
                    // Flow Control frame received
                    uint8_t flow_status = rx_message.data[0] & 0x0F;
                    uint8_t block_size = rx_message.data[1];
                    uint8_t st_min = rx_message.data[2];

                    if (flow_status == FLOW_STATUS_CTS)
                    {
                        // Continue to Send (CTS)
                        fc_params->block_size = block_size;
                        fc_params->st_min = st_min;
                        ESP_LOGW(TAG, "Continue to Send (CTS), block_size: %u, st_min: %u", fc_params->block_size, fc_params->st_min);
                        return FRAME_99_ACK;
                    }
                    else if (flow_status == FLOW_STATUS_WT)
                    {
                        // Wait (WT), continue waiting with remaining time
                        ESP_LOGE(TAG, "Flow Control: Wait requested by receiver");
                        continue;
                    }
                    else if (flow_status == FLOW_STATUS_OVFLW)
                    {
                        // Overflow (OVFLW)
                        ESP_LOGE(TAG, "Flow Control: Receiver overflowed");
                        return FRAME_99_OVFLW;
                    }
                    else
                    {
                        // Unsupported flow status
                        ESP_LOGE(TAG, "Flow Control: Unsupported flow status 0x%02X", flow_status);
                        return FRAME_99_UFC;
                    }
                }
                // Not a Flow Control frame; continue waiting with remaining time
                continue;
            }
            // Not matching identifier or frame type; continue waiting with remaining time
            continue;
        }
        // Queue receive timeout occurred
        else if (remaining_time == 0)
        {
            break;
        }
    }

    ESP_LOGE(TAG, "Flow Control frame not received within timeout");
    return FRAME_99_FC_TIMEOUT;
}

```

**Comment by yannick9906** [comment]
Hey @meatpiHQ thank you for that quick reply. I increased the stack size and used your function from above. Now all responses get received and the full transfer is being sent. Then I noticed that the first transfer takes longer than my timeout for the frame_99 ACK (5 sec), now it seems to work as it should.
